File: benchmarks/dblp_opt.cpp
#include <map>
#include <chrono>

#include "Context.hpp"

using namespace Centaurus;

int placeholder;

void *parseDocument(const SymbolContext<char>& ctx)
{
  return ctx.value<int>(1);
}

static std::vector<std::string*> result;

void *parseDblpRoot(const SymbolContext<char>& ctx)
{
  for (int i = 0; i < ctx.count(); i++) {
    result.emplace_back(ctx.value<std::string>(i+1));
  }
  return &result;
}

void *parseArticle(const SymbolContext<char>& ctx)
{
  return (ctx.count() > 0) ? new std::string(ctx.start(), ctx.len()) : nullptr;
}

void *parseYearInfo(const SymbolContext<char>& ctx)
{
  return (ctx.count() > 0) ? &placeholder : nullptr;
}

void *parseTargetYearInfo(const SymbolContext<char>& ctx)
{
  return &placeholder;
}

int main(int argc, const char *argv[])
{
    if (argc < 1) return 1;
    int worker_num = std::atoi(argv[1]);
    bool no_action = argc >= 3 && argv[2] == std::string("dry");
    bool size = argc >= 3 && argv[2] == std::string("size");
    bool debug = argc >= 3 && argv[2] == std::string("debug");

    const char *input_path = "datasets/dblp.xml";
    const char *grammar_path = "grammar/dblp.cgr";

    Context<char> context{grammar_path};

    if (!no_action) {
      context.attach(L"Document", parseDocument);
      context.attach(L"DblpRoot", parseDblpRoot);
      context.attach(L"Article", parseArticle);
      context.attach(L"YearInfo", parseYearInfo);
      context.attach(L"TargetYearInfo", parseTargetYearInfo);
    }

    using namespace std::chrono;

    auto start = high_resolution_clock::now();;

    context.parse(input_path, worker_num);

    auto end = high_resolution_clock::now();;

    std::cout << worker_num << " " << duration_cast<milliseconds>(end - start).count() << std::endl;

    if (size) {
      std::cout << result.size() << std::endl;
    }
    if (debug) {
      for (auto p : result) {
        std::cout << *p << std::endl;
      }
    }
    return 0;
}

File: PythonApp1/dblp_opt.py
#!/usr/bin/env python3

import sys
import os
import logging
import multiprocessing as mp
import time

os.environ['CENTAURUS_DL_PATH'] = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'build')

from centaurus import *

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

class DBLPListener(object):
    def parseDocument(self, ctx):
        return ctx[0]
    def parseDblpRoot(self, ctx):
        return list(ctx)
    def parseArticle(self, ctx):
        if ctx:
            return ctx.read()
    def parseYearInfo(self, ctx):
        return True if ctx else None
    def parseTargetYearInfo(self, ctx):
        return True
    def defaultact(self, ctx):
        pass

if __name__ == '__main__':
    app_name = os.path.basename(sys.argv[0]).split('.')[0]
    num_workers = int(sys.argv[1]) if len(sys.argv) > 1 else 1
    logging.basicConfig(filename='debug_{}-p{}.log'.format(app_name, num_workers), filemode='w', level=logging.DEBUG)

    context = Context('../grammar/dblp.cgr')
    input_path = '../datasets/dblp.xml'
    listener = DBLPListener()
    context.attach(listener)
    context.start(num_workers)
    start_time = time.time()
    result = context.parse(input_path)
    end_time = time.time()
    context.stop()
    print(num_workers, end_time - start_time)

    if len(sys.argv) > 2 and sys.argv[2].lower() == 'size':
        print(len(result))
    if len(sys.argv) > 2 and sys.argv[2].lower() == 'debug':
        with open('result_{}.log'.format(app_name), 'w') as dst:
            for s in result:
                print(s)
